### DISCUSSION TITLE
wasm apps jump around when interacting

### DISCUSSION THREAD

**Discussion by g-braeunlich** [emilk/egui#746] [Q&A]
After migration of egui from 0.10 -> 0.14, I now have a new behavior.
Every time, I click on an app mounted to a `<canvas>` element, this element gets assigned `position: absolute; top: 0%;` and thus jumps to the top.
Is there a way to prevent that?

Example: https://g-braeunlich.gitlab.io/egui_extras/widgets/index.html

**Comment by emilk** [comment]
I believe this is caused by this line: https://github.com/emilk/egui/blob/master/egui_web/src/lib.rs#L1206

To handle text input on mobile, an invisible text input field is moved to under the cursor (if there is a text field under the mouse). To make sure a mobile keyboard can fit, the ill-named function `manipulate_agent` moves up the the egui canvas to make room for it without covering the text field. When there is no text input, the canvas is restored to full screen - which is the assumed size of an egui app.

Basically: egui_web assumes the egui canvas takes up the full height of the screen, and so breaks for your use case.

I think embedding egui into a larger html document like you do is very cool and an important use case to support, but I haven't had the time to work on it.

I'm actually a bit surprised that it works at all, considering is a function `resize_canvas_to_screen_size` which resizes the egui canvas to the full screen (so why aren't you egui elements taking up the whole screen? I don't know).

Anyway, the fix should be to check whether we are in the fullscreen or embedded mode, and only move the canvas if the former. `egui_web` could check when initialized (read some property of the canvas?) or perhaps pass that in by argument to [`start`](https://docs.rs/egui_web/0.14.1/egui_web/backend/fn.start.html)?

Would you have time to take a look at it?

Fixing text input for the embedded use case i probably a bit more work.

**Reply by g-braeunlich** [reply]
Hey @emilk
Thx for your fast answer! I am willing to give it a try! Still could you give me some more info about fullscreen mode? Are you referring to the fullscreen mode in a browser or is this some egui internal state (in this case: where would that be stored / accessable)?

**Reply by g-braeunlich** [reply]
Also if you are fine with it: I would turn this discussion into an issue.

**Reply by emilk** [reply]
>I am willing to give it a try!

Awesome, thanks!

> Still could you give me some more info about fullscreen mode? Are you referring to the fullscreen mode in a browser or is this some egui internal state

I am referring to when the egui canvas fills the whole browser document height, as it does in the egui web demo, and in egui_template. "fullscreen" is a bit misleading, but I didn't find a better word for it.

**Reply by g-braeunlich** [reply]
I just discovered a workaround:
If you wrap the `<canvas>` into a `<div style="position: relative;">`, then the `top: 0` will relate to the div and the canvas wont jump.
On the other side I also could confirm your assumption: Commenting out the lines you mentioned above also prevents jumping.

Do you still want to change the behavior?
I saw that in the demo, the `<canvas>` has `<body>` as the direct parent. So that condition could be used to determine whether to reposition.
Alternatively one also could define and use an egui specific attribute for the canvas to decide.
But I also would see now the status as is as an option (just using the workaround).
What do you think?

**Reply by g-braeunlich** [reply]
Correction: this is only a partial workaround. The elements around the div still behave weirdly (with position: absolute, no space is allocated for the element).

**Reply by g-braeunlich** [reply]
Another suggestion: We could decide if we want to reposition based on the `position: absolute` and `top: 0%` being set in the first place. Then you cannot break anything. But it would also mean, that you would have to change your html for the demo

**Reply by g-braeunlich** [reply]
Do you have a preference for any of those suggestions? Or do you have an other idea?

**Reply by emilk** [reply]
I'm not the best person to judge this - perhaps @n2 has some better input. I'm generally pretty crap when it comes to all things web.

We ideally don't want to break peoples web apps when they update to a newer `egui_web`, so the ideal fix would be one that doesn't need changes to index.html. But if a simple solution is also preferred, and if it makes the egui_web code a lot simpler than a change in index.html is fine.

**Comment by n2** [comment]
Moving canvas to absolute position of 0 is based on the assumptions:

1. there is only one canvas, and
2. that canvas is located at height of 0, and
3. there is no other html elements(text, image, etc.) in page,

 which is what currently happens in index.html. 
As for your example, none of these three assumptions meet.

Based on the 3 assumptions, commenting out the lines, or `wrap the <canvas> into a <div style="position: relative;">` may also result to wierd jumping on mobile devices. 
I am not good at web programing, the reason why I move canvas only is because I don't know how to move the whole html document that contains the canvas. If you have a better idea, please improve it @g-braeunlich.
